### ISSUE TITLE
Attempted import error

### ISSUE THREAD

**Issue by rquigley** [dolanmiu/docx#2365]
I'm seeing the following error when used within a Next.js app ([test repo here](https://github.com/rquigley/bug-next-docx)).

```
Attempted import error: 'Document' is not exported from 'docx' (imported as 'Document').
Attempted import error: 'Paragraph' is not exported from 'docx' (imported as 'Paragraph').
Attempted import error: 'TextRun' is not exported from 'docx' (imported as 'TextRun').
```

I tried downgrading to 8.2.2 and still run into the same issue. This can be worked around by using CJS ([see here](https://github.com/rquigley/bug-next-docx/blob/main/app/page.tsx#L7)).

Thank you so much for your effort. This is a great library.

**Comment by dolanmiu** [comment]
ESM and CJS seems to always cause issues, will need to look into it

**Comment by dustin-rcg** [comment]
Perhaps because the package.json has specifically exported the mjs version for `import`, but the commonjs version for `require`?

https://github.com/dolanmiu/docx/blob/3eb98533aea49540d4f7d5b39d1828a4baeb22aa/package.json#L9C1-L16C7

```
    "exports": {
        ".": {
            "require": "./build/index.cjs",
            "types": "./build/index.d.ts",
            "import": "./build/index.mjs",
            "default": "./build/index.mjs"
        }
    },
```

Do you want your users to be forced to switch to ESM?

**Comment by dolanmiu** [comment]
@dustin-rcg This is done on purpose so that both ESM and CJS are both supported

Would prefer if both CJS and ESM are both supported to be honest

**Comment by dustin-rcg** [comment]
@dolanmiu Unfortunately, in order for us to use this package, we must use the following in tsconfig:

```
    "module": "commonjs",
    "moduleResolution": "node",
```

The note on that moduleResolution setting says `Deprecated, use "Node10" in TypeScript 5.0+ instead`. And the note on "node10" says `It’s recommended to use "Node16" instead`.  But we cannot start to use Node16 or even NodeNext, because these module resolutions force `module` to match, and this totally breaks `docx` for us.  So we are forced to use these [archaic deprecated settings](https://www.typescriptlang.org/tsconfig#moduleResolution) to consume this package. There has to be a better way.

It's 2023 and we're supposed to be using NodeNext now: `This is the recommended setting for libraries and Node.js applications`. We are up to nodejs v20+ and typescript v5+.

> 'node16' or 'nodenext' for modern versions of Node.js. Node.js v12 and later supports both ECMAScript imports and CommonJS require, which resolve using different algorithms. These moduleResolution values, when combined with the corresponding [module](https://www.typescriptlang.org/tsconfig#module) values, picks the right algorithm for each resolution based on whether Node.js will see an import or require in the output JavaScript code.

> 'node10' (previously called 'node') for Node.js versions older than v10, which only support CommonJS require. You probably won’t need to use node10 in modern code.

So based on this, it seems the way docx is published must be incorrect, since NodeNext should work with both CJS and MJS, but _it fails to work with docx currently_. If we try to use NodeNext as the tsconfig docs tell us, docx blows up:

> The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("docx")' call instead.
  To convert this file to an ECMAScript module, add the field `"type": "module"` to '/Users/dustin/code/repos/RCG-Sirius/async-workers/package.json'.ts(1479)

And then when I try to convert to ESM, I still get the error that nothing is exported from `docx` except for `Document`:

> Module '"docx"' has no exported member 'Packer'.ts(2305)

This happens for _every_ import from `docx` except for `Document`; so neither CJS docx nor MJS docx works with typescript in the proper settings. It's really sad because we just redid a number of our reports using this package, and we like it a lot, but now it becomes another package that forces ESM on us (for NodeNext) and even then fails to compile due to missing types. We are using yarn pnp, so I don't know if that also contributes to the problem, but _not_ using pnp is not an option, because we depend on modules to be hoisted from shared packages in our monorepo to avoid multiple instances from separate `node_modules` breaking decorator metadata and `instanceof`.

**Comment by kalda341** [comment]
@dustin-rcg I'm also encountering this issue. Not specific to yarn pnp as we're using pnpm.

**Comment by rquigley** [comment]
Fixed with 8.3.0. Thanks so much!

**Comment by tothdanielax** [comment]
Not next.js related, but the import error is still open for me when I try to import patchDocument in ESM based Node.js code.

```ts
// TS2305: Module "docx" has no exported member patchDocument 
import {patchDocument} from "docx";
```

tsconfig.json:
```json
{
  "compilerOptions": {
    /* Env */
    "target": "ES2022",
    "module": "Node16",
    "moduleResolution": "Node16",
    "lib": [
      "ES2023"
    ],
    "outDir": "dist",
    /* Settings  */
    "esModuleInterop": true,
    "useDefineForClassFields": true,
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "allowImportingTsExtensions": false,
    "removeComments": true,
    "inlineSourceMap": true,
    "verbatimModuleSyntax": true,
    "moduleDetection": "force",
    "noEmit": true,
    /* Linting */
    "allowUnreachableCode": false,
    "allowUnusedLabels": true,
    "exactOptionalPropertyTypes": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": false,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
  }
}
```

Environment:

- ts: v5.2.2
- node: v21.2.0
- pnpm: v8.11.0
- macos 14.2.1
